sprint_13

Gvozdkov opened the pull request.

Commit:
Вынесена вся лишняя ответственность из вьюконтроллера в презентер и создано для каждой ответственности свою сущность в коде.

Commit:
Написаны тесты WebViewTests

Commit:
Написан UI тест для авторизации в приложении, тест успешно работает.

Commit:
Добавлено 2 UI теста:
1. Работа с лентой
2. Работа с профилем

Commit:
Проведён рефакторинг ImagesList

Commit:
1. Написан Unit test для ImagesList
2. Добавлена иконка приложения

helecta commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -33,7 +37,10 @@ final class AuthViewController: UIViewController {
 extension AuthViewController: WebViewViewControllerDelegate {
     func webViewViewController(_ vc: WebViewViewController, didAuthenticateWithCode code: String) {
-        delegate?.authViewController(self, didAuthenticateWithCode: code)
+        dismiss(animated: true) { [weak self] in
```
У тебя успеет показаться экран авторизации.

Как вариант решения:
1. Добавить в UIBlockingProgressHUD свойство `isShowing`:
```
final class UIBlockingProgressHUD {
    
    static var isShowing: Bool = false
    
    private static var window: UIWindow? {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let window = windowScene.windows.first
        else {
            return nil
        }
        return window
    }

    static func show() {
        isShowing = true
        window?.isUserInteractionEnabled = false
        ProgressHUD.show()
    }

    static func dismiss() {
        isShowing = false
        window?.isUserInteractionEnabled = true
        ProgressHUD.dismiss()
    }
}
```

2. Во viewDidAppear добавить проверку:
```
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        guard UIBlockingProgressHUD.isShowing == false else { return }
        if let token = OAuth2TokenStorage.shared.token {
            fetchProfile(with: token)
            switchToTabBarController()
        } else {
            guard let authViewController = UIStoryboard(name: "Main", bundle: .main).instantiateViewController(withIdentifier: "AuthViewController") as? AuthViewController else {
                assertionFailure("Failed to show Authentication Screen")
                return
            }

            authViewController.delegate = self
            authViewController.modalPresentationStyle = .fullScreen
            present(authViewController, animated: true)
        }
    }
```

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -7,39 +7,46 @@ protocol WebViewViewControllerDelegate: AnyObject {
     @IBOutlet private var webView: WKWebView!
     @IBOutlet private var progressView: UIProgressView!
     private var estimatedProgressObservation: NSKeyValueObservation?
     weak var delegate: WebViewViewControllerDelegate?
```
Лучше delegate перенести в презентер, т.к это больше про бизнес-логику

helecta replied:
Не забывай организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

Подробнее почитать можно тут: [Организация кода](https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization)

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -7,39 +7,46 @@ protocol WebViewViewControllerDelegate: AnyObject {
-        progressView.progress = Float(webView.estimatedProgress)
-        progressView.isHidden = fabs(webView.estimatedProgress - 1.0) <= 0.0001
+    func setProgressValue(_ newValue: Float) {
+        progressView.progress = newValue
```
если использовать progressView.setProgress(…), то можно сделать заполнение прогресс бара более плавным.

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -7,39 +7,46 @@ protocol WebViewViewControllerDelegate: AnyObject {
+        webView.load(request)
     }
     
     @IBAction private func didTapBackButton(_ sender: Any) {
```
Этот метод должен перенаправлять в презентер, уже ответственность презентера решать что делать дальше. В данном случае перенаправлять делегату

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -59,32 +66,23 @@ extension WebViewViewController: WKNavigationDelegate {
             decisionHandler(.allow)
```
Здесь не видно код, но лучше код выше перенаправлять в презентер. Т.е тут должно быть так:
```
    func webView(
        _ webView: WKWebView,
        decidePolicyFor navigationAction: WKNavigationAction,
        decisionHandler: @escaping (WKNavigationActionPolicy) -> Void
    ) {
        if let url = navigationAction.request.url,
           let code = presenter?.code(from: url) {
            presenter?.didAuthenticateWithCode(code)
            decisionHandler(.cancel)
        } else {
            decisionHandler(.allow)
        }
```

helecta commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -1,44 +1,54 @@
-    
+
     private lazy var dateFormatter: DateFormatter = {
         let formatter = DateFormatter()
```
Если так настроить дата форматтер:
```
private lazy var dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "dd MMMM yyyy"
        formatter.locale = Locale(identifier: "ru_RU")
        return formatter
    }()
```
То будет выглядеть как в макете, без  `г.`

helecta commented on ImageFeedUITests/ImageFeedUITests.swift:
```diff
@@ -0,0 +1,86 @@
+        XCTAssertTrue(webView.waitForExistence(timeout: 5))
+        loginTextField.tap()
+        loginTextField.typeText(login)
+        webView.swipeUp()
```
Как вариант еще можно нажимать на кнопку Done на клавиатуре:
```
if app.keyboards.element(boundBy: 0).exists {
            app.toolbars.buttons["Done"].tap()
        }
```

Commit:
Исправлено:
1. После ввода логина и пароля раньше успевал показываться экран  авторизации. Теперь после ввода личных данных переход идет сразу в ленту с фотографиями.
2. Изменен формат даты под фотографией (было: 12.12.2023 г., стало:  12.12.2023).
3. У UIBlockingProgressHUD изменен визуальный стиль теперь он соответствует макету.
